File: 1143. Longest Common Subsequence-3.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n = len(text1)
        m = len(text2)

        f = [0] * (m + 1)

        for i, x in enumerate(text1):
            pre = f[0]  # 0
            for j, y in enumerate(text2):
                temp = f[j + 1]
                f[j + 1] = pre + 1 if x == y else max(f[j + 1], f[j])
                pre = temp

        return f[m]

File: 1143. Longest Common Subsequence-3.cpp
class Solution {
public:
  vector<vector<int>> memo;
  int longestCommonSubsequence(string text1, string text2) {
    int m = text1.size(), n = text2.size();
    memo.assign(m, vector<int>(n, -1));
    return dfs(text1, text2, m - 1, n - 1);
  }

  int dfs(string &text1, string &text2, int i, int j) {
    if (i < 0 || j < 0)
      return 0;
    int &res = memo[i][j];
    if (res != -1)
      return res;
    if (text1[i] == text2[j])
      return res = dfs(text1, text2, i - 1, j - 1) + 1;
    return res = max(dfs(text1, text2, i - 1, j), dfs(text1, text2, i, j - 1));
  }
};
